Bridge as a gz sim system [rolling]

caguero opened the pull request:
# 🎉 New feature

## Summary
<!--Explain changes made, the expected behavior, and provide any other additional
context (e.g., screenshots, gifs) if appropriate.-->

This PR runs the ROS<->Gz bridge as a gz sim system. It's targeting ROS 2 rolling. I'll create a separate PR with some updates that we discussed offline.

## Test it
<!--Explain how reviewers can test this new feature manually.-->

```
ros2 launch ros_gz_sim_demos ros_gz.launch.py
```

Verify the output to confirm that two topics are bridged:
```
[ruby $(which ign) gazebo-1] [INFO] [1708634732.070359776] [ros_gz_bridge]: Creating ROS->GZ Bridge: [ros_chatter (std_msgs/msg/String) -> gz_chatter (ignition.msgs.StringMsg)] (Lazy 1)
[ruby $(which ign) gazebo-1] [INFO] [1708634732.071384449] [ros_gz_bridge]: Creating GZ->ROS Bridge: [gz_chatter (ignition.msgs.StringMsg) -> ros_chatter (std_msgs/msg/String)] (Lazy 0)
```

Then, you should see a ROS topic `/ros_chatter`:

```
caguero@cold:~/ros_gz_ws$ ros2 topic list
/parameter_events
/ros_chatter
/rosout
```

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

luca-della-vedova committed:
Update README to include harmonic (#473)

* Update README to include harmonic

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>

* Remove humble-harmonic combination

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>

---------

Signed-off-by: Luca Della Vedova <lucadv@intrinsic.ai>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

Commit:
Add tip about non-parallel builds to README (#477)

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

Commit:
Additional tip for limiting number of cpus (#480)

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

Commit:
Undeprecate use of commandline flags (#491)

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

kscottz committed:
Update README.md to clarify what will be supported in Jazzy (#499)

Also Added a note pointing to REP-2000 for additional details.
---------

Signed-off-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>
Co-authored-by: Addisu Z. Taddese <addisu@openrobotics.org>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

Commit:
Add a virtual destructor to suppress compiler warning (#502)

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

Commit:
Correctly export ros_gz_bridge for downstream targets (#503)

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

caguero committed:
Feedback updates.

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

caguero committed:
Adding files.

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

caguero commented:
@ahcorde , do you know if these tests are supposed to compile? It seems a problem sourcing `setup.bash` in `rolling`.

ahcorde commented on ros_gz_sim/include/ros_gz_sim/ros_gz_bridge_system.hpp:
```diff
@@ -0,0 +1,57 @@
+
+  /// \brief Destructor.
+  ~ROSGzBridgeSystem() override;
+
+  // Documentation inherited.
+  void Configure(
+    const gz::sim::Entity & _entity,
+    const std::shared_ptr<const sdf::Element> & _sdf,
+    gz::sim::EntityComponentManager & _ecm,
+    gz::sim::EventManager & _eventMgr) override;
```
add includes for these headers

caguero replied:
Added in [83bf13e](https://github.com/gazebosim/ros_gz/pull/511/commits/83bf13e8308622df22d0ba4d03dee3817d10cbf3)

caguero committed:
Includes

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

ahcorde reviewed (approved):
@azeey or @mjcarroll do you mind to take a look too ?

caguero committed:
Use resource_retriever in the bridge (#515)

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>
Signed-off-by: Carlos Agüero <cen.aguero@gmail.com>
Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

ahcorde commented on ros_gz_sim/src/ros_gz_bridge_system.cpp:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  if (!_sdf->HasElement("config_file")) {
+    std::cerr << "No <config_file> found. Plugin disabled." << std::endl;
```
as this is a Gazebo plugin, should we use here `gzerr` ? remove `<iostream>` header too

caguero replied:
Changed in [78067c4](https://github.com/gazebosim/ros_gz/pull/511/commits/78067c419f272ce86ec1a3060929f4fd037afabe).

caguero committed:
Tweak

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

ahcorde committed:
ign to gz (#519)

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

ahcorde committed:
Remove iron CI from rolling (#518)

Signed-off-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

Commit:
Keep rolling on Jammy and apply a temporary fix for rosdep (#522)

Pins rosdep to a known good version based on https://wiki.ros.org/SnapshotRepository

---------

Signed-off-by: Addisu Z. Taddese <addisu@openrobotics.org>
Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

azeey commented on ros_gz_bridge/src/bridge_config.cpp:
```diff
@@ -173,31 +174,30 @@ std::vector<BridgeConfig> readFromYaml(std::istream & in)
 {
   std::vector<BridgeConfig> ret;
-  std::ifstream in(filename);
+  resource_retriever::Retriever r;
```
nit: single letter variable names violate our style guide.  Can you use a more descriptive variable?

caguero replied:
Renamed in [13f4008](https://github.com/gazebosim/ros_gz/pull/511/commits/13f40087d99b92c5bc4c8b157a951aad1efc0d9f)

azeey commented on ros_gz_bridge/include/ros_gz_bridge/ros_gz_bridge.hpp:
```diff
@@ -68,6 +68,9 @@ class RosGzBridge : public rclcpp::Node
   rclcpp::TimerBase::SharedPtr heartbeat_timer_;
+
+  /// \brief Whether the config_file parameter was parsed or not.
+  bool config_file_parsed_;
```
nit: This can be initialized here and not in the constructor
```suggestion
  bool config_file_parsed_{false};
```

caguero replied:
Updated in [13f4008](https://github.com/gazebosim/ros_gz/pull/511/commits/13f40087d99b92c5bc4c8b157a951aad1efc0d9f)

azeey commented on ros_gz_bridge/test/bridge_config.cpp:
```diff
@@ -70,15 +70,16 @@ class BridgeConfig : public ::testing::Test
 {
-  auto results = ros_gz_bridge::readFromYamlFile("test/config/minimum.yaml");
+  auto results = ros_gz_bridge::readFromYamlFile(
+    "package://ros_gz_bridge/test/config/minimum.yaml");
```
Can we keep at least one test that uses the old file path (I.e, just `test/config/minimum.yaml`)  so we are sure we're not breaking behavior?

caguero replied:
Sure. Changed in [13f4008](https://github.com/gazebosim/ros_gz/pull/511/commits/13f40087d99b92c5bc4c8b157a951aad1efc0d9f).

azeey commented on ros_gz_sim/include/ros_gz_sim/ros_gz_bridge_system.hpp:
This file does not need to be installed. It can live in `ros_gz_sim/src`.

caguero replied:
Moved in [13f4008](https://github.com/gazebosim/ros_gz/pull/511/commits/13f40087d99b92c5bc4c8b157a951aad1efc0d9f)

azeey commented on ros_gz_sim/src/ros_gz_bridge_system.cpp:
```diff
@@ -0,0 +1,99 @@
+    rclcpp::NodeOptions().append_parameter_override("config_file", filename));
+
+  // Create the executor.
+  this->dataPtr->exec =
+    std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
```
The [documentation](https://docs.ros.org/en/humble/Concepts/Intermediate/About-Executors.html#types-of-executors) on Multi-Threaded Executors seems to indicate we need to use callback groups to take advantage of parallelism. @mjcarroll Can you confirm?

azeey commented:
Can you also fix DCO?

caguero committed:
Tweaks

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

caguero committed:
ifdef

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

caguero committed:
Merge

Signed-off-by: Carlos Agüero <caguero@openrobotics.org>

caguero commented:
> Can you also fix DCO?

Fixed

azeey commented:
@caguero Do we want to close this one now that the composable nodes PRs are merged?
